File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/core/services/conversation_store.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';
import 'package:runanywhere/runanywhere_protos.dart' as proto;

/// File-based persistence for conversation history.
///
/// Mirrors iOS `Core/Services/ConversationStore.swift`: conversations are
/// created in-memory, become persistent on the first message, and every
/// update is written back to a `Conversations/<id>.json` file.
class ConversationStore extends ChangeNotifier {
  static final ConversationStore shared = ConversationStore._();

  ConversationStore._() {
    unawaited(_ready);
  }

  static const _defaultTitle = 'New Chat';
  static const _encoder = JsonEncoder.withIndent('  ');

  List<Conversation> _conversations = [];
  Conversation? _currentConversation;
  Directory? _conversationsDirectory;
  late final Future<void> _ready = _initialize();

  List<Conversation> get conversations => _conversations;
  Conversation? get currentConversation => _currentConversation;

  Future<void> _initialize() async {
    final documentsDir = await getApplicationDocumentsDirectory();
    _conversationsDirectory = Directory('${documentsDir.path}/Conversations');

    if (!await _conversationsDirectory!.exists()) {
      await _conversationsDirectory!.create(recursive: true);
    }

    await loadConversations();
  }

  /// Create a new conversation. Not added to the list or written to disk
  /// until the first message arrives (iOS parity).
  Conversation createConversation({String? title}) {
    final now = DateTime.now();
    final conversation = Conversation(
      id: now.microsecondsSinceEpoch.toString(),
      title: title ?? _defaultTitle,
      createdAt: now,
      updatedAt: now,
      messages: const [],
    );
    _currentConversation = conversation;
    return conversation;
  }

  /// Insert or update [conversation], persist it to disk, and return the
  /// stored copy (with a refreshed `updatedAt`).
  Conversation updateConversation(Conversation conversation) {
    final updated = conversation.copyWith(updatedAt: DateTime.now());

    final index = _conversations.indexWhere((c) => c.id == updated.id);
    if (index >= 0) {
      _conversations[index] = updated;
    } else {
      // First time this conversation is persisted (first message sent).
      _conversations.insert(0, updated);
    }

    if (_currentConversation?.id == updated.id) {
      _currentConversation = updated;
    }

    unawaited(_saveConversation(updated));
    notifyListeners();
    return updated;
  }

  /// Append [message] to [conversation], derive a title from the first user
  /// message if still untitled, persist, and return the updated conversation.
  ///
  /// iOS additionally generates "smart titles" via Apple FoundationModels —
  /// that path is Apple-platform-gated, so Flutter keeps the deterministic
  /// fallback title only.
  Conversation addMessage(Message message, Conversation conversation) {
    var updated = conversation.copyWith(
      messages: [...conversation.messages, message],
    );

    if (updated.title == _defaultTitle) {
      final firstUserContent = updated.messages
          .firstWhere(
            (m) =>
                m.role == proto.MessageRole.MESSAGE_ROLE_USER &&
                m.content.trim().isNotEmpty,
            orElse: () => message,
          )
          .content;
      if (firstUserContent.trim().isNotEmpty) {
        updated = updated.copyWith(title: _generateTitle(firstUserContent));
      }
    }

    return updateConversation(updated);
  }

  /// Delete a conversation
  void deleteConversation(Conversation conversation) {
    _conversations.removeWhere((c) => c.id == conversation.id);
    if (_currentConversation?.id == conversation.id) {
      _currentConversation = null;
    }
    unawaited(_deleteConversationFile(conversation.id));
    notifyListeners();
  }

  /// Load all conversations from disk
  Future<void> loadConversations() async {
    if (_conversationsDirectory == null) return;

    try {
      final files = _conversationsDirectory!.listSync();
      final loadedConversations = <Conversation>[];

      for (final file in files) {
        if (file is File && file.path.endsWith('.json')) {
          try {
            final content = await file.readAsString();
            final json = jsonDecode(content) as Map<String, dynamic>;
            loadedConversations.add(Conversation.fromJson(json));
          } catch (e) {
            debugPrint('Error loading conversation: $e');
          }
        }
      }

      _conversations = loadedConversations
        ..sort((a, b) => b.updatedAt.compareTo(a.updatedAt));
      notifyListeners();
    } catch (e) {
      debugPrint('Error loading conversations: $e');
    }
  }

  Future<void> _saveConversation(Conversation conversation) async {
    await _ready;
    try {
      final file = File('${_conversationsDirectory!.path}/${conversation.id}.json');
      await file.writeAsString(_encoder.convert(conversation.toJson()));
    } catch (e) {
      debugPrint('Error saving conversation: $e');
    }
  }

  Future<void> _deleteConversationFile(String id) async {
    await _ready;

    try {
      final file = File('${_conversationsDirectory!.path}/$id.json');
      if (await file.exists()) {
        await file.delete();
      }
    } catch (e) {
      debugPrint('Error deleting conversation file: $e');
    }
  }

  /// Deterministic title fallback: first line of the first user message,
  /// capped at 50 characters (mirrors iOS `generateTitle(from:)`).
  String _generateTitle(String content) {
    const maxLength = 50;
    final cleaned = content.trim();
    final firstLine = cleaned.split('\n').first;
    return firstLine.length > maxLength
        ? firstLine.substring(0, maxLength)
        : firstLine;
  }
}

/// Conversation model
class Conversation {
  final String id;
  final String title;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<Message> messages;
  final String? modelName;
  final String? frameworkName;

  const Conversation({
    required this.id,
    required this.title,
    required this.createdAt,
    required this.updatedAt,
    required this.messages,
    this.modelName,
    this.frameworkName,
  });

  String get lastMessagePreview {
    if (messages.isEmpty) return 'Start a conversation';

    final lastMessage = messages.last;
    final preview = lastMessage.content.trim().replaceAll('\n', ' ');

    return preview.length > 100 ? preview.substring(0, 100) : preview;
  }

  Conversation copyWith({
    String? title,
    DateTime? updatedAt,
    List<Message>? messages,
    String? modelName,
    String? frameworkName,
  }) {
    return Conversation(
      id: id,
      title: title ?? this.title,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      messages: messages ?? this.messages,
      modelName: modelName ?? this.modelName,
      frameworkName: frameworkName ?? this.frameworkName,
    );
  }

  factory Conversation.fromJson(Map<String, dynamic> json) => Conversation(
        id: json['id'] as String,
        title: json['title'] as String,
        createdAt: DateTime.parse(json['createdAt'] as String),
        updatedAt: DateTime.parse(json['updatedAt'] as String),
        messages: (json['messages'] as List<dynamic>)
            .map((m) => Message.fromJson(m as Map<String, dynamic>))
            .toList(),
        modelName: json['modelName'] as String?,
        frameworkName: json['frameworkName'] as String?,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
        'messages': messages.map((m) => m.toJson()).toList(),
        if (modelName != null) 'modelName': modelName,
        if (frameworkName != null) 'frameworkName': frameworkName,
      };
}

/// Message model
class Message {
  final String id;
  final proto.MessageRole role;
  final String content;
  final String? thinkingContent;
  final DateTime timestamp;
  final MessageAnalytics? analytics;

  const Message({
    required this.id,
    required this.role,
    required this.content,
    this.thinkingContent,
    required this.timestamp,
    this.analytics,
  });

  Message copyWith({
    String? id,
    proto.MessageRole? role,
    String? content,
    String? thinkingContent,
    DateTime? timestamp,
    MessageAnalytics? analytics,
  }) {
    return Message(
      id: id ?? this.id,
      role: role ?? this.role,
      content: content ?? this.content,
      thinkingContent: thinkingContent ?? this.thinkingContent,
      timestamp: timestamp ?? this.timestamp,
      analytics: analytics ?? this.analytics,
    );
  }

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        id: json['id'] as String,
        role: _messageRoleFromJson(json['role'] as String?),
        content: json['content'] as String,
        thinkingContent: json['thinkingContent'] as String?,
        timestamp: DateTime.parse(json['timestamp'] as String),
        analytics: json['analytics'] != null
            ? MessageAnalytics.fromJson(
                json['analytics'] as Map<String, dynamic>)
            : null,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'role': role.name,
        'content': content,
        if (thinkingContent != null) 'thinkingContent': thinkingContent,
        'timestamp': timestamp.toIso8601String(),
        if (analytics != null) 'analytics': analytics!.toJson(),
      };
}

/// Message analytics for tracking generation metrics
class MessageAnalytics {
  final String messageId;
  final String? modelName;
  final String? framework;
  final double? timeToFirstToken;
  final double? totalGenerationTime;
  final int inputTokens;
  final int outputTokens;
  final double? tokensPerSecond;
  final bool wasThinkingMode;
  final proto.ChatMessageStatus completionStatus;

  const MessageAnalytics({
    required this.messageId,
    this.modelName,
    this.framework,
    this.timeToFirstToken,
    this.totalGenerationTime,
    this.inputTokens = 0,
    this.outputTokens = 0,
    this.tokensPerSecond,
    this.wasThinkingMode = false,
    this.completionStatus =
        proto.ChatMessageStatus.CHAT_MESSAGE_STATUS_COMPLETE,
  });

  factory MessageAnalytics.fromJson(Map<String, dynamic> json) =>
      MessageAnalytics(
        messageId: json['messageId'] as String,
        modelName: json['modelName'] as String?,
        framework: json['framework'] as String?,
        timeToFirstToken: json['timeToFirstToken'] as double?,
        totalGenerationTime: json['totalGenerationTime'] as double?,
        inputTokens: json['inputTokens'] as int? ?? 0,
        outputTokens: json['outputTokens'] as int? ?? 0,
        tokensPerSecond: json['tokensPerSecond'] as double?,
        wasThinkingMode: json['wasThinkingMode'] as bool? ?? false,
        completionStatus:
            _chatMessageStatusFromJson(json['completionStatus'] as String?),
      );

  Map<String, dynamic> toJson() => {
        'messageId': messageId,
        if (modelName != null) 'modelName': modelName,
        if (framework != null) 'framework': framework,
        if (timeToFirstToken != null) 'timeToFirstToken': timeToFirstToken,
        if (totalGenerationTime != null)
          'totalGenerationTime': totalGenerationTime,
        'inputTokens': inputTokens,
        'outputTokens': outputTokens,
        if (tokensPerSecond != null) 'tokensPerSecond': tokensPerSecond,
        'wasThinkingMode': wasThinkingMode,
        'completionStatus': completionStatus.name,
      };
}

proto.MessageRole _messageRoleFromJson(String? value) {
  switch (value) {
    case 'MESSAGE_ROLE_SYSTEM':
    case 'system':
      return proto.MessageRole.MESSAGE_ROLE_SYSTEM;
    case 'MESSAGE_ROLE_ASSISTANT':
    case 'assistant':
      return proto.MessageRole.MESSAGE_ROLE_ASSISTANT;
    case 'MESSAGE_ROLE_USER':
    case 'user':
      return proto.MessageRole.MESSAGE_ROLE_USER;
    default:
      return proto.MessageRole.MESSAGE_ROLE_USER;
  }
}

proto.ChatMessageStatus _chatMessageStatusFromJson(String? value) {
  switch (value) {
    case 'CHAT_MESSAGE_STATUS_PENDING':
      return proto.ChatMessageStatus.CHAT_MESSAGE_STATUS_PENDING;
    case 'CHAT_MESSAGE_STATUS_STREAMING':
      return proto.ChatMessageStatus.CHAT_MESSAGE_STATUS_STREAMING;
    case 'CHAT_MESSAGE_STATUS_FAILED':
    case 'failed':
    case 'timeout':
      return proto.ChatMessageStatus.CHAT_MESSAGE_STATUS_FAILED;
    case 'CHAT_MESSAGE_STATUS_CANCELLED':
    case 'interrupted':
      return proto.ChatMessageStatus.CHAT_MESSAGE_STATUS_CANCELLED;
    case 'CHAT_MESSAGE_STATUS_COMPLETE':
    case 'complete':
    default:
      return proto.ChatMessageStatus.CHAT_MESSAGE_STATUS_COMPLETE;
  }
}

File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Core/Services/ConversationStore.swift
import Combine
import Foundation
import RunAnywhere
#if canImport(FoundationModels)
import FoundationModels
#endif

// Note: Message, MessageAnalytics and ConversationAnalytics are now in separate model files

// MARK: - Conversation Store

@MainActor
class ConversationStore: ObservableObject {
    static let shared = ConversationStore()

    @Published var conversations: [Conversation] = []
    @Published var currentConversation: Conversation?

    private let documentsDirectory: URL

    private static func getDocumentsDirectory() -> URL {
        guard let url = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            fatalError("Unable to access documents directory")
        }
        return url
    }
    private let conversationsDirectory: URL
    private let attachmentsDirectory: URL
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private init() {
        documentsDirectory = Self.getDocumentsDirectory()
        conversationsDirectory = documentsDirectory.appendingPathComponent("Conversations")
        attachmentsDirectory = conversationsDirectory.appendingPathComponent("Attachments")

        // Create conversations directory if it doesn't exist
        try? FileManager.default.createDirectory(at: conversationsDirectory, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: attachmentsDirectory, withIntermediateDirectories: true)

        // Set up encoder/decoder
        encoder.outputFormatting = .prettyPrinted
        encoder.dateEncodingStrategy = .iso8601
        decoder.dateDecodingStrategy = .iso8601

        // Load existing conversations
        loadConversations()
    }

    // MARK: - Public Methods

    func createConversation(title: String? = nil) -> Conversation {
        let conversation = Conversation(
            id: UUID().uuidString,
            title: title ?? "New Chat",
            createdAt: Date(),
            updatedAt: Date(),
            messages: [],
            modelName: nil,
            frameworkName: nil
        )

        // Don't add to conversations list yet - wait until first message is added
        currentConversation = conversation
        // Don't save empty conversation - wait until first message is added

        return conversation
    }

    func updateConversation(_ conversation: Conversation) {
        var updated = conversation
        updated.updatedAt = Date()

        if let index = conversations.firstIndex(where: { $0.id == conversation.id }) {
            // Update existing conversation
            conversations[index] = updated
        } else {
            // First time adding this conversation (when first message is sent)
            conversations.insert(updated, at: 0)
        }

        if currentConversation?.id == conversation.id {
            currentConversation = updated
        }

        saveConversation(updated)
    }

    func deleteConversation(_ conversation: Conversation) {
        conversations.removeAll { $0.id == conversation.id }

        if currentConversation?.id == conversation.id {
            currentConversation = conversations.first
        }

        // Delete file
        let fileURL = conversationFileURL(for: conversation.id)
        try? FileManager.default.removeItem(at: fileURL)
        try? FileManager.default.removeItem(at: attachmentDirectory(for: conversation.id))
    }

    func addMessage(_ message: Message, to conversation: Conversation) {
        var updated = conversation
        updated.messages.append(message)
        updated.updatedAt = Date()

        // Always try to generate a fallback title if still "New Chat"
        if updated.title == "New Chat" {
            if let firstUserMessage = updated.messages.first(where: { $0.role == .user }),
               !firstUserMessage.content.isEmpty {
                updated.title = generateTitle(from: firstUserMessage.content)
            }
        }

        updateConversation(updated)

        // Try to generate smart title with Foundation Models after first AI response
        if message.role == .assistant && updated.messages.count >= 2 {
            let conversationId = updated.id
            Task { @MainActor in
                await self.generateSmartTitleIfNeeded(for: conversationId)
            }
        }
    }

    func saveAttachment(
        data: Data,
        filename: String,
        kind: MessageAttachment.Kind,
        conversationID: String,
        detail: String? = nil,
        previewText: String? = nil
    ) throws -> MessageAttachment {
        let directory = attachmentDirectory(for: conversationID)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)

        let storedFilename = "\(UUID().uuidString)-\(safeAttachmentFilename(filename))"
        let fileURL = directory.appendingPathComponent(storedFilename)
        try data.write(to: fileURL, options: [.atomic])

        return MessageAttachment(
            kind: kind,
            filename: filename,
            detail: detail,
            relativePath: "Conversations/Attachments/\(conversationID)/\(storedFilename)",
            previewText: previewText
        )
    }

    // MARK: - Foundation Models Title Generation

    /// Public method to generate smart title for a conversation
    func generateSmartTitleForConversation(_ conversationId: String) async {
        await generateSmartTitleIfNeeded(for: conversationId)
    }

    private func generateSmartTitleIfNeeded(for conversationId: String) async {
        #if canImport(FoundationModels)
        guard #available(iOS 26.0, macOS 26.0, *) else { return }

        // Find the conversation
        guard let conversation = conversations.first(where: { $0.id == conversationId }) else {
            return
        }

        // Get the fallback title to compare
        let firstUserMessage = conversation.messages.first { $0.role == .user }
        let fallbackTitle = firstUserMessage.map { generateTitle(from: $0.content) } ?? "New Chat"

        // Only generate if title is still the default or fallback
        let currentTitle = conversation.title
        guard currentTitle == "New Chat" || currentTitle == fallbackTitle else {
            return
        }

        // Use the SDK's platform capability gate so simulator and unsupported
        // devices keep the deterministic fallback title.
        guard SystemFoundationModels.isAvailable else { return }

        // Create conversation text from first few messages
        let conversationText = conversation.messages
            .prefix(4)
            .map { msg in
                "\(msg.role == .user ? "User" : "Assistant"): \(msg.content.prefix(200))"
            }
            .joined(separator: "\n")

        do {
            let titleSession = LanguageModelSession(
                instructions: Instructions("""
                    You are an expert at creating descriptive, readable chat titles.
                    Generate a clear title (2-5 words) that captures the main topic.
                    Respond in the same language as the conversation.
                    Only output the title, nothing else.
                    """)
            )

            let titlePrompt = """
            Create a descriptive, readable title for this conversation:

            \(conversationText)

            Title:
            """

            let response = try await titleSession.respond(to: Prompt(titlePrompt))
            let title = response.content
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .replacingOccurrences(of: "\"", with: "")

            // Update the conversation with the AI-generated title
            if !title.isEmpty, var conv = self.conversations.first(where: { $0.id == conversationId }) {
                conv.title = String(title.prefix(50))
                self.updateConversation(conv)
            }
        } catch {
            // Keep the fallback title
        }
        #endif
    }

    func loadConversation(_ id: String) -> Conversation? {
        if let conversation = conversations.first(where: { $0.id == id }) {
            currentConversation = conversation
            return conversation
        }

        // Try to load from disk
        let fileURL = conversationFileURL(for: id)
        if let data = try? Data(contentsOf: fileURL),
           let conversation = try? decoder.decode(Conversation.self, from: data) {
            conversations.append(conversation)
            currentConversation = conversation
            return conversation
        }

        return nil
    }

    // MARK: - Search

    func searchConversations(query: String) -> [Conversation] {
        guard !query.isEmpty else { return conversations }

        let lowercasedQuery = query.lowercased()

        return conversations.filter { conversation in
            // Search in title
            if conversation.title.lowercased().contains(lowercasedQuery) {
                return true
            }

            // Search in messages
            return conversation.messages.contains { message in
                message.content.lowercased().contains(lowercasedQuery)
            }
        }
    }

    // MARK: - Private Methods

    private func loadConversations() {
        do {
            let files = try FileManager.default.contentsOfDirectory(
                at: conversationsDirectory,
                includingPropertiesForKeys: [.contentModificationDateKey]
            )

            var loadedConversations: [Conversation] = []

            for file in files where file.pathExtension == "json" {
                if let data = try? Data(contentsOf: file),
                   let conversation = try? decoder.decode(Conversation.self, from: data) {
                    loadedConversations.append(conversation)
                }
            }

            // Sort by update date, newest first
            conversations = loadedConversations.sorted { $0.updatedAt > $1.updatedAt }

            // Don't automatically set current conversation - let ChatViewModel create a new one
        } catch {
            print("Error loading conversations: \(error)")
        }
    }

    private func saveConversation(_ conversation: Conversation) {
        let fileURL = conversationFileURL(for: conversation.id)

        do {
            let data = try encoder.encode(conversation)
            try data.write(to: fileURL)
        } catch {
            print("Error saving conversation: \(error)")
        }
    }

    private func conversationFileURL(for id: String) -> URL {
        conversationsDirectory.appendingPathComponent("\(id).json")
    }

    private func attachmentDirectory(for id: String) -> URL {
        attachmentsDirectory.appendingPathComponent(id)
    }

    private func safeAttachmentFilename(_ filename: String) -> String {
        let cleaned = filename
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .components(separatedBy: CharacterSet(charactersIn: "/:\\?%*|\"<>"))
            .joined(separator: "-")
        return cleaned.isEmpty ? "attachment" : cleaned
    }

    private func generateTitle(from content: String) -> String {
        // Take first 50 characters or up to first newline
        let maxLength = 50
        let cleaned = content.trimmingCharacters(in: .whitespacesAndNewlines)

        if let newlineIndex = cleaned.firstIndex(of: "\n") {
            let firstLine = String(cleaned[..<newlineIndex])
            return String(firstLine.prefix(maxLength))
        }

        return String(cleaned.prefix(maxLength))
    }
}

// MARK: - Conversation Model

struct Conversation: Identifiable, Codable {
    let id: String
    var title: String
    let createdAt: Date
    var updatedAt: Date
    var messages: [Message]
    var modelName: String?
    var frameworkName: String?

    // NEW: Conversation-level analytics
    var analytics: ConversationAnalytics?
    var performanceSummary: PerformanceSummary?

    var summary: String {
        guard !messages.isEmpty else { return "No messages" }

        let messageCount = messages.count
        let userMessages = messages.filter { $0.role == .user }.count
        let assistantMessages = messages.filter { $0.role == .assistant }.count

        return "\(messageCount) messages • \(userMessages) from you, \(assistantMessages) from AI"
    }

    var lastMessagePreview: String {
        guard let lastMessage = messages.last else { return "Start a conversation" }

        let preview = lastMessage.content
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "\n", with: " ")

        return String(preview.prefix(100))
    }
}

// Performance summary for quick access
struct PerformanceSummary: Codable {
    let averageResponseTime: TimeInterval
    let totalTokens: Int
    let mainModel: String?
    let completionRate: Double
    let averageTokensPerSecond: Double

    init(from messages: [Message]) {
        let analyticsMessages = messages.compactMap { $0.analytics }

        if !analyticsMessages.isEmpty {
            let count = Double(analyticsMessages.count)
            let totalTime = analyticsMessages.compactMap { $0.totalGenerationTime }.reduce(0, +)
            averageResponseTime = totalTime / count
            totalTokens = analyticsMessages.reduce(0) { $0 + $1.inputTokens + $1.outputTokens }
            mainModel = analyticsMessages.first?.modelName
            let completed = analyticsMessages.filter { $0.completionStatus == .complete }.count
            completionRate = Double(completed) / count
            let totalTPS = analyticsMessages.compactMap { $0.averageTokensPerSecond }.reduce(0, +)
            averageTokensPerSecond = totalTPS / count
        } else {
            averageResponseTime = 0
            totalTokens = 0
            mainModel = nil
            completionRate = 0
            averageTokensPerSecond = 0
        }
    }
}
